TheMovieDb/Network/APIClient.swift

```diff
@@ -0,0 +1,66 @@
+//
+//  APIClient.swift
+//  TheMovieDb
+//
+//  Created by Rob Cruz on 18/11/21.
+//
+
+import Foundation
+
+protocol APIClient {
+    var session: URLSession { get }
+    func fetch<T: Decodable>(with request: URLRequest, decode: @escaping (Decodable) -> T?, completion: @escaping (Result<T, APIError>) -> Void)
+}
+
+// MARK: - JSON Generic Data Task
+
+extension APIClient {
+    typealias JSONTaskCompletionHandler = (Decodable?, APIError?) -> Void
+    private func decodingTask<T: Decodable>(with request: URLRequest, decodingType: T.Type, completionHandler completion: @escaping JSONTaskCompletionHandler) -> URLSessionDataTask {
+        let task = session.dataTask(with: request) { data, response, _ in
+            guard let httpResponse = response as? HTTPURLResponse else {
+                completion(nil, .requestFailed)
+                return
+            }
+            if httpResponse.statusCode == 200 {
+                guard let data = data else { return completion(nil, .invalidData) }
+                    do {
+                        let jsonData = JSONDecoder()
+                        jsonData.keyDecodingStrategy = .convertFromSnakeCase
+                        let genericModel = try jsonData.decode(decodingType, from: data)
```

Comment:
Is there a case where the decodingType differs from the Decodable generic type? Can you use `.decode(T.self, from: data)` ?